### ISSUE TITLE
Decide on the precedence of type ascription

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#7304]
Does it bind tightly or loosely?

How do we parse: `not true : Bool` does the ascription apply to `true` or to `not true`?

**Comment by mlutze** [comment]
What do other languages do?

**Comment by JonathanStarup** [comment]
Another option - remove inline ascription. require let-binding if you want to ascribe

**Comment by magnus-madsen** [comment]
@herluf-ba Time for you to shine :)

**Comment by magnus-madsen** [comment]
> What do other languages do?

What other languages? We only came up with Scala and a quick search did not reveal anything.

**Comment by magnus-madsen** [comment]
> Another option - remove inline ascription. require let-binding if you want to ascribe

I think thats too strict. We want e.g. `Nil: List[Int32]`

**Comment by mlutze** [comment]
> > What do other languages do?
> 
> What other languages? We only came up with Scala and a quick search did not reveal anything.

OCaml requires parentheses: https://v2.ocaml.org/manual/expr.html
So it has to be ` (Nil : List[Int32])`.

Haskell binds to a large expression.
```
Prelude> 1 == 1 :: Integer

<interactive>:3:1: error:
    • Couldn't match expected type ‘Integer’ with actual type ‘Bool’
    • In the expression: 1 == 1 :: Integer
      In an equation for ‘it’: it = 1 == 1 :: Integer
```
but
```
Prelude> 1 == 1 :: Bool
True
```

**Comment by magnus-madsen** [comment]
I would lean towards requiring parens.

**Comment by herluf-ba** [comment]
Typescript does "type-assertions" (different syntax, same thing)
```typescript
"str1" == "str2" as boolean
// This means:
"str1" == ("str2" as boolean)
```
So ascriptions bind tight here.

I don't think there is a silver bullet. Using a loose binding means you have to use parens to ascribe a small sub-expression, while using a tight bind means using parens for ascribing larger expressions made of multiple sub-expressions. Either way around the programmer is inconvenienced in one case and happy in the other.

To me the whole issue arises since we humans have little intuition about the precedence of the ascription operator! 
No one is confused about `1 + 2 * 4`. We know `*` binds tighter than `+`. But if we do `1 + 3 : Int32 * 4` then it is not clear anymore.
That is why I kind of like requiring parens too, because that would disambiguate all cases.

**Comment by JonathanStarup** [comment]
I agree with parens, but unfortunately, being consistent and clear means that we force the worst syntax all of the time 😄

But maybe its fine. In my code i'd always use let-ascriptions anyway. When do you even use `Nil: List[Int32]` when you're not forced to by scala type inference?

**Comment by magnus-madsen** [comment]
Let us do require parens. Its forward compatible, i.e. we can always make the rule less strict. 

@herluf-ba If you implement it in the new parser, and point out where we have to fix code in the std lib then we can get someone to do it :)

**Comment by magnus-madsen** [comment]
@jaschdoc We need all uses of ascription to have parenthesis.

For example, if we had

```
def foo(): List[Int32] = Nil : List[Int32]
```

This now has to become:

```
def foo(): List[Int32] = (Nil : List[Int32]) // notice the paren.
```

To find out where to make the changes, I would edit:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Weeder.scala#L1281

and print the source location (sp2).

IN ADDITION, please check if the documentation on doc.flix.dev has to be updated. Thanks! 👍

**Comment by magnus-madsen** [comment]
@herluf-ba Ascription should now be fixed. Please confirm :)

**Comment by magnus-madsen** [comment]
@jaschdoc Remember to check the documentation. Maybe its OK?

https://doc.flix.dev/type-ascriptions.html#type-ascriptions

**Comment by magnus-madsen** [comment]
Awesome!
